Data-handling CAFMaker script(s)

brucehoward-physics opened the pull request:
Add fhicl that is basically a copy of `cafmakerjob_icarus.fcl` but nullifies a some labels and a list of labels that belong to the simulation:

> physics.producers.cafmaker.G4Label: ""
> physics.producers.cafmaker.GenLabel: ""
> physics.producers.cafmaker.SimChannelLabel: ""
> physics.producers.cafmaker.SystWeightLabels: []

and makes it use a `caf_ana_data_sequence` which is the same as `caf_ana_sequence` removing the `mcreco` instance.

Tested that it ran over a few data events in v09_37 and produced a CAF file with some of the reco branches filled that I'd expect to be filled for data.

I'm tagging Gray and Chris B as reviewers here, and in fact have a question for each:
* Chris B: the resulting CAF file had many (33?) `globalTree` s -- is that expected?
* Gray: are there preliminary data calo constants that we should point this fcl too, or if not should we maybe comment that as a `TODO`?

cjbackhouse commented:
There should only be one globalTree per se. ROOT does this strange auto-save thing that can make trees show up multiple times in a browser, but I don't think that's too important. Did you maybe run over 33 input runs? How many entries does the globalTree have?

cjbackhouse reviewed (approved):
Looks good. I think there is probably some fat that can be trimmed.

cjbackhouse commented:
Fix for the multiple globalTrees in https://github.com/SBNSoftware/sbncode/pull/206

brucehoward-physics commented on fcl/caf/cafmaker_defs.fcl:
```diff
@@ -193,6 +193,18 @@ caf_ana_sequence: [ mcreco,
     # TODO: rns??
     ]
 
+caf_ana_data_sequence: [
```
I think @gputnam is possibly the originator of this naming scheme, but I think the name is probably because these are some of the additional modules that are run in the CAFMaker fcl but _before_ the `cafmaker` process itself. That said I see the potential raised here for ambiguity between `caf_ana_sequence` and `CAFAna`.

Gray should comment if my reasoning for the title is incorrect or if he is not the originator of the name schema.

gputnam replied:
Yeah that's correct @brucehoward-physics. I don't have strong opinions about the name so feel free to change it to whatever.

brucehoward-physics commented:
Ah I missed the last comment with a proposed fix for the multiple Global Trees. Thanks for the fast turn-around.

gputnam commented on fcl/caf/cafmaker_defs.fcl:
```diff
@@ -193,6 +193,18 @@ caf_ana_sequence: [ mcreco,
 
+caf_ana_data_sequence: [
+    # Calorimetry
+    pandoraCaloGausCryoE, pandoraCaloGausCryoW,
```
Because calorimetry is not ready, my suggestion would be to exclude the Calorimetry / Low-Energy-Proton-Reco steps from the data chain.

brucehoward-physics replied:
I don't necessarily agree we want to turn off the calorimetry outputs, even if the calorimetry constants aren't quite ready. Wouldn't it be better to have this info for testing how things look and just somehow note for the record that these aren't fully finalized? Even if you remove them, calorimetry info will exist for some things, e.g. the showers.

gputnam commented on fcl/caf/cafmakerjob_icarus_data.fcl:
```diff
@@ -0,0 +1,56 @@
+services:
+{
+  ParticleInventoryService:  @local::standard_particleinventoryservice
+  BackTrackerService:        @local::standard_backtrackerservice
```
Following up -- I don't believe we should need the BackTracker. I would suggest removing this and fixing any code that fails of BackTracker or ParticleInventory are not present.

brucehoward-physics replied:
Commented below that this was just missed, my bad. I suspect it "should just work" but we'll see.

gputnam reviewed (changes requested):
I would suggest changing the list of analysis modules.

brucehoward-physics commented:
As ever "it should just work" are famous last words... In fact both services seem to be needed by one of the modules (not yet sure which).

brucehoward-physics commented:
The requirement for backtracker and particle inventory is baked into CAFMaker --

https://github.com/SBNSoftware/sbncode/blob/d466a8e336fa3ae950ca33ab6309e5aa3ca8e7d5/sbncode/CAFMaker/CAFMaker_module.cc#L753-L754

brucehoward-physics commented:
Tried to address most of the comments in the latest commit. Didn't change the name of **other** `caf_ana_sequence` instances other than the data one here. I don't even know what the feelings about changing those now are, as the workflow is already being tested.

Still to address:

1. Whether to remove all of the calo-focused preprocessing in the data sequence. I removed the vertex charge/stub, but am not so sure we should remove the calo/chi2pid (instead suggested we make sure it's announced that the current production will not have finalized values here.)

2. Removing the ParticleInventory and BackTracker servicers (see [Issue in SBNCode](https://github.com/SBNSoftware/sbncode/issues/210) and [related PR...](https://github.com/SBNSoftware/sbncode/pull/209) ). If those are accepted then I have to add another commit here.

And one other thing: it seems there is a CAFMaker param for `FillTrueParticle` - do we want to explicitly set that to false here?

brucehoward-physics commented:
**Context**

Per discussion at the SBN Analysis Frameworks meeting I believe we have agreement on **not** turning off the extra reco pieces for data and acknowledgement that we should somehow inform analyzers that some of these products (calorimetry etc.) that this isn't in it's final form for data.

Also per discussion at the meeting, the `FillTrueParticle` option I asked about above can be left as-is.

This PR now touches many more `cafmakerjob` fhicls, changing instances of `caf_ana_` to `caf_preprocess_` to try to make it less ambiguous with `CAFAna`. I tested that `cafmakerjob_icarus_sce_genie_and_fluxwgt.fcl` at least worked as part of testing, though I didn't check every one of the fhicls.

I removed the comment about defining modules from the `cafmakerjob_icarus.fcl` as well, but left the out stream _here_ since that actually is used in the `nocaf` fhicl. I removed it from the `data` fhicl as there is no such link-up there, at least at the moment.

**Especially for reviewers**

I think that's basically it, except the following question: Do we also want to remove this from the cafmakerjob_icarus_data.fcl?
`NuRandomService: @local::per_event_NuRandomService`

Other than this, I _think_ this is now in the hands of the reviewers @cjbackhouse and @gputnam to comment further (especially if I missed addressing something)/request further things as necessary. Thanks!

**For Release Managers (Please take note!)**

This PR now depends on https://github.com/SBNSoftware/sbncode/pull/209

SFBayLaser commented:
trigger build

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/13900&builds=)

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for c7:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/13901&builds=)

FNALbuild commented:
:warning: CI build for ICARUS Warning at phase ci_tests ICARUS on slf7 for c7:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/3310&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/3310&platform=Linux%203.10.0-1160.45.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20c7:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13901&builds=)

SFBayLaser commented:
trigger build

FNALbuild commented:
:heavy_check_mark: CI build for LArSoft Succeeded on slf7 for e20:prof -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app//view_builds/index?offset=0&builds=lar_ci/13903&builds=)

FNALbuild commented:
:warning: CI build for ICARUS Warning at phase ci_tests ICARUS on slf7 for e20:prof - ignored warnings for build -- details available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/view_builds/index?offset=0&builds=icarus_ci/3312&builds=)

:rotating_light: For more details about the warning phase, check the [ci_tests ICARUS phase logs](https://dbweb0.fnal.gov/LarCI/app/ns:ICARUS/build_detail/phase_details?build_id=icarus_ci/3312&platform=Linux%203.10.0-1160.45.1.el7.x86_64&phase=ci_tests&buildtype=slf7%20e20:prof)

parent CI build details are available through the [CI dashboard](https://dbweb0.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/13903&builds=)
